### PR TITLE
Changes in TensorFlow for C++ implementation of AC

### PR DESCRIPTION
**Pull request by shreyajain17**
The initialization for the model was earlier taken from MXNet and now is taken from C++ in CoreML format. The code is being changed so the C++ implementation for AC toolkit can call tensorflow for training.

### CODE REVIEW

File: src/python/turicreate/toolkits/activity_classifier/_tf_model_architecture.py [hunk 1]
```diff
@@ -112,46 +126,75 @@ def __init__(self, net_params, batch_size, num_features, num_classes, prediction
...
         init = _tf.compat.v1.global_variables_initializer()
         self.sess.run(init)
-
+        
         # Assign the initialised weights from MXNet to tensorflow 
         for key in net_params.keys():
             if key in weights.keys():
                 if key.startswith('conv'):
-                    self.sess.run(_tf.assign(_tf.get_default_graph().get_tensor_by_name(key+":0"), _np.transpose(net_params[key].asnumpy(), (2, 1, 0))))
+                    net_params[key] = _np.reshape(_np.transpose(net_params[key], (3, 1, 0, 2)), (prediction_window, self.num_features, _net_params['conv_h']))
```
Comment (nickjong) [inline src/python/turicreate/toolkits/activity_classifier/_tf_model_architecture.py]: Could we factor our this conversion into a utility function that can be used by other toolkits?

File: src/python/turicreate/toolkits/activity_classifier/_tf_model_architecture.py [hunk 2]
```diff
@@ -112,46 +126,75 @@ def __init__(self, net_params, batch_size, num_features, num_classes, prediction
...
+            probabilities = _np.reshape(prob, (prob.shape[0], prob.shape[1]*prob.shape[2]))
+            result = {'loss' : _np.array(loss),  'output': probabilities }
+        
         return result
 
 
     def export_weights(self):
 
-        # Export weights in Core ML Spec
+        # Export weights in Core ML Spec
```
Comment (nickjong) [inline src/python/turicreate/toolkits/activity_classifier/_tf_model_architecture.py]: Please refrain from adding trailing whitespace. Does clang-format or other tool work on Python code?

File: src/python/turicreate/toolkits/activity_classifier/_tf_model_architecture.py (hunk 1 above)
Comment (shreyajain17) [reply src/python/turicreate/toolkits/activity_classifier/_tf_model_architecture.py]: But this is 1D Convs and like OD/DC use 2D Convs. They would all be named conv0_weight/conv1_weight or just conv_weight!

File: src/python/turicreate/toolkits/activity_classifier/_tf_model_architecture.py (hunk 2 above)
Comment (shreyajain17) [reply src/python/turicreate/toolkits/activity_classifier/_tf_model_architecture.py]: I don't think it works on python code

File: src/python/turicreate/toolkits/activity_classifier/_tf_model_architecture.py [hunk 3]
```diff
@@ -6,41 +6,48 @@
...
 
 class ActivityTensorFlowModel(TensorFlowModel):
     
     def __init__(self, net_params, batch_size, num_features, num_classes, prediction_window, seq_len):
 
+        # Supresses verbosity to only errors
+        _tf.compat.v1.logging.set_verbosity(_tf.compat.v1.logging.ERROR)
+
+
+        _net_params = {
```
[the remark this answers is not in the document]
Comment (shreyajain17) [reply src/python/turicreate/toolkits/activity_classifier/_tf_model_architecture.py]: Will change it

**Comment by shreyajain17** [comment]
After fixing the calculation for stride, accuracy seems to be fixed. Loss seems similar while we train but it is a bit higher in validation while using tensorflow
```Using CPU to create model
+--------------+----------------+--------------+---------------------+-----------------+--------------+
| Iteration    | Train Accuracy | Train Loss   | Validation Accuracy | Validation Loss | Elapsed Time |
+--------------+----------------+--------------+---------------------+-----------------+--------------+
| 1            | 0.653397       | 0.938731     | 0.738333            | 1.4709          | 17.50s       |
| 2            | 0.833786       | 0.485992     | 0.872412            | 1.29395         | 19.31s       |
| 3            | 0.861107       | 0.394308     | 0.903071            | 1.16795         | 21.25s       |
| 4            | 0.881622       | 0.328445     | 0.921922            | 1.05883         | 22.94s       |
| 5            | 0.890961       | 0.291353     | 0.929587            | 0.977964        | 24.77s       |
| 6            | 0.903998       | 0.25555      | 0.922178            | 0.832293        | 26.41s       |
| 10           | 0.914019       | 0.204149     | 0.932575            | 0.460567        | 33.92s       |
+--------------+----------------+--------------+---------------------+-----------------+--------------+```

File: src/ml/neural_net/CMakeLists.txt [hunk 4]
```diff
@@ -62,7 +62,13 @@ make_library(tctensorflow SHARED
   tf_model_backend.cpp
 )
 
-set_target_properties(tctensorflow PROPERTIES SUFFIX ".so")
+set(TF_BUILD_DIR ${CMAKE_BINARY_DIR}/src/python/turicreate/toolkits)
```
Comment (nickjong) [inline src/ml/neural_net/CMakeLists.txt]: What is this for?

File: src/ml/neural_net/tf_model_backend.cpp [hunk 5]
```diff
@@ -74,7 +72,30 @@ PYBIND11_MODULE(libtctensorflow, m) {
       });
 }
 
-tf_model_backend::tf_model_backend(pybind11::object model): model_(model) {}
+
+tf_model_backend::tf_model_backend(int n, int c_in, int h_in, int w_in, int c_out, int h_out, int w_out,
+    const float_array_map& config, const float_array_map& weights, std::string model_name) {
+
+  if (model_name == "activity_classifier") {
```
Comment (nickjong) [inline src/ml/neural_net/tf_model_backend.cpp]: I don't think it makes sense to handle AC specific code here. I would either go back to the way you had it before, or go all the way forward to passing just the model name and all the other parameters (including batch size, input shape, config, weights) in a dictionary to a generic (non-AC-specific) Python module that handles everything AC-specific.

Offhand, for now I would go back to the way we had it before. I think it'll be more work than we want to do now to make this clean.

File: src/ml/neural_net/tf_compute_context.hpp [hunk 6]
```diff
@@ -42,6 +42,9 @@ class tf_compute_context: public compute_context {
 
   std::unique_ptr<image_augmenter> create_image_augmenter(
       const image_augmenter::options &opts) override;
+
```
Comment (nickjong) [inline src/ml/neural_net/tf_compute_context.hpp]: What is this whitespace for?

File: src/python/turicreate/toolkits/_tf_utils.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+	in TensorFlow.
+
+	All the weights from the C++ implementation are four dimensional.
+
+	"""
+	return np.array(array, copy=False)
+
+def convert_conv1d_coreml_to_tf(conv_weights):
+	"""
+	The Convolutional weights in CoreML specification is sent as [C, H, 1, W]
```
Comment (nickjong) [inline src/python/turicreate/toolkits/_tf_utils.py]: `[C, H, 1, W]` doesn't look right to me. 

From the `NeuralNetwork.proto` spec:
```
    /**
     * Weights associated with this layer.
     * If convolution (``isDeconvolution == false``), weights have the shape
     * ``[outputChannels, kernelChannels, kernelHeight, kernelWidth]``, where kernelChannels == inputChannels / nGroups
     * If deconvolution (``isDeconvolution == true``) weights have the shape
     * ``[kernelChannels, outputChannels / nGroups, kernelHeight, kernelWidth]``, where kernelChannels == inputChannels
     */
```

File: src/python/turicreate/toolkits/_tf_utils.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+	"""
+	i2h = np.concatenate((i2h_i, i2h_c, i2h_f, i2h_o), axis=0)
+	h2h = np.concatenate((h2h_i, h2h_c, h2h_f, h2h_o), axis=0)
+	lstm = np.concatenate((i2h, h2h), axis=1)
+	return np.transpose(lstm)
+
+def convert_dense_coreml_to_tf(dense_weights):
+	"""
+	The Dense layer weights from CoreML are [C_in, C_out, 1, 1] and need to be
```
Comment (nickjong) [inline src/python/turicreate/toolkits/_tf_utils.py]:
```
message InnerProductLayerParams {
    uint64 inputChannels = 1; /// Input size: C_in.
    uint64 outputChannels = 2; /// Output size: C_out.

    bool hasBias = 10; /// Whether a bias is added or not.

    WeightParams weights = 20; /// Weight matrix [C_out, C_in].
    WeightParams bias = 21; /// Bias vector [C_out].
}
```

Or from TC code:
```
void wrap_network_params(const std::string& name,
                         const InnerProductLayerParams& inner_product,
                         float_array_map* params_out) {

  const size_t n = inner_product.outputchannels();
  const size_t c = inner_product.inputchannels();

  shared_float_array weights = weight_params_float_array::create_view(
      {n, c, 1, 1}, inner_product.weights());
  params_out->emplace(name + "_weight", std::move(weights));
```

File: src/python/turicreate/toolkits/activity_classifier/_activity_classifier.py [hunk 9]
```diff
@@ -29,7 +29,7 @@
 from .util import random_split_by_session as _random_split_by_session
 from .util import _MIN_NUM_SESSIONS_FOR_SPLIT
 
-
+import turicreate.toolkits.libtctensorflow
```
Comment (nickjong) [inline src/python/turicreate/toolkits/activity_classifier/_activity_classifier.py]: We don't want to import TF by default, only if the user has opted into using the TF beta implementation

File: src/python/turicreate/toolkits/activity_classifier/_tf_model_architecture.py [hunk 10]
```diff
@@ -101,40 +98,38 @@ def __init__(self, net_params, batch_size, num_features, num_classes, prediction
         out = _tf.add(_tf.matmul(dense, weights['dense1_weight']), biases['dense1_bias'])
         out = _tf.reshape(out, (-1, self.seq_len, self.num_classes))
         self.probs = _tf.nn.softmax(out)
-
-        # Weights
+
```
Comment (nickjong) [inline src/python/turicreate/toolkits/activity_classifier/_tf_model_architecture.py]: Please avoid adding trailing whitespace

File: src/python/turicreate/toolkits/activity_classifier/_tf_model_architecture.py [hunk 11]
```diff
@@ -101,40 +98,38 @@ def __init__(self, net_params, batch_size, num_features, num_classes, prediction
...
+        # Assign the initialised weights from the C++ implementation to tensorflow 
         for key in net_params.keys():
             if key in weights.keys():
                 if key.startswith('conv'):
-                    net_params[key] = _np.reshape(_np.transpose(net_params[key], (3, 1, 0, 2)), (prediction_window, self.num_features, _net_params['conv_h']))
+                    net_params[key] = _utils.convert_conv1d_coreml_to_tf(net_params[key])
                     self.sess.run(_tf.assign(_tf.get_default_graph().get_tensor_by_name(key+":0"), net_params[key]))
                 elif key.startswith('dense'):
-                    net_params[key] =  _np.reshape(_np.transpose(net_params[key], (1, 0, 2, 3)), (net_params[key].shape[1], net_params[key].shape[0]))
+                    # net_params[key] =  _np.reshape(_np.transpose(net_params[key], (1, 0, 2, 3)), (net_params[key].shape[1], net_params[key].shape[0]))
```
Comment (nickjong) [inline src/python/turicreate/toolkits/activity_classifier/_tf_model_architecture.py]: Can we remove this commented-out code?

File: src/ml/neural_net/CMakeLists.txt (hunk 4 above)
Comment (shreyajain17) [reply src/ml/neural_net/CMakeLists.txt]: The shared library was being made in `turicreate/debug/src/ml/neural_net`. This lines moves it to `turicreate/debug/src/python/turicreate/toolkits`. Then I can simply import it in the python file as `import turicreate.toolkits.libtctensorflow`.

File: src/ml/neural_net/tf_model_backend.cpp (hunk 5 above)
Comment (shreyajain17) [reply src/ml/neural_net/tf_model_backend.cpp]: Changed this

File: src/python/turicreate/toolkits/activity_classifier/_tf_model_architecture.py (hunk 11 above)
Comment (shreyajain17) [reply src/python/turicreate/toolkits/activity_classifier/_tf_model_architecture.py]: Yes

File: src/python/turicreate/toolkits/activity_classifier/_tf_model_architecture.py (hunk 10 above)
Comment (shreyajain17) [reply src/python/turicreate/toolkits/activity_classifier/_tf_model_architecture.py]: Yeah

File: src/python/turicreate/toolkits/activity_classifier/_activity_classifier.py (hunk 9 above)
Comment (shreyajain17) [reply src/python/turicreate/toolkits/activity_classifier/_activity_classifier.py]: Moved it inside the block where user sets it to `True`

File: src/python/turicreate/toolkits/activity_classifier/_tf_model_architecture.py [hunk 12]
```diff
@@ -28,7 +28,7 @@ def __init__(self, net_params, batch_size, num_features, num_classes, prediction
 
          # Suppresses verbosity to only errors
         _tf.compat.v1.logging.set_verbosity(_tf.compat.v1.logging.ERROR)
-
+
```
Comment (nickjong) [inline src/python/turicreate/toolkits/activity_classifier/_tf_model_architecture.py]: Please avoid introducing trailing whitespace into the codebase
